File: python/day2.py
OPP_ROCK = "A"
OPP_PAPER = "B"
OPP_SCISSORS = "C"

MY_ROCK = "X"
MY_PAPER = "Y"
MY_SCISSORS = "Z"

MY_LOSS = "X"
MY_DRAW = "Y"
MY_WIN = "Z"


ROCK = 1
PAPER = 2
SCISSORS = 3


def parse_choice(s: str) -> int:
    if s == OPP_ROCK or s == MY_ROCK:
        return ROCK
    elif s == OPP_PAPER or s == MY_PAPER:
        return PAPER
    elif s == OPP_SCISSORS or s == MY_SCISSORS:
        return SCISSORS
    else:
        raise ValueError


LOSE_SCORE = 0
DRAW_SCORE = 3
WIN_SCORE = 6


def p1_score_round(opp: str, my: str) -> int:
    opp = parse_choice(opp)
    my = parse_choice(my)
    return p1_score_round_parsed(opp, my)


def p1_score_round_parsed(opp: int, my: int) -> int:
    if opp == my:
        return my + DRAW_SCORE
    elif (my - opp) % 3 == 1:
        return my + WIN_SCORE
    else:
        return my + LOSE_SCORE


DRAW = 0
WIN = 1
LOSS = 2


def parse_outcome(s: str) -> int:
    if s == MY_WIN:
        return WIN
    elif s == MY_LOSS:
        return LOSS
    elif s == MY_DRAW:
        return DRAW
    else:
        raise ValueError


def cause_outcome(opp: int, desired: int) -> int:
    if desired == DRAW:
        return opp
    elif desired == WIN:
        return max((opp + 1) % 4, 1)
    else:
        return SCISSORS if opp == ROCK else (opp + 2) % 3


def p2_score_round(opp: str, desired: str) -> int:
    opp_choice = parse_choice(opp)
    desired_out = parse_outcome(desired)
    my_choice = cause_outcome(opp_choice, desired_out)
    assert 1 <= my_choice <= 3, (opp, desired, my_choice)
    return p1_score_round_parsed(opp_choice, my_choice)


if __name__ == '__main__':
    with open("../day2_input.txt") as f:
        p1_sum = 0
        p2_sum = 0
        for line in f:
            first, second = line.strip().split()
            p1_sum += p1_score_round(first, second)
            p2_sum += p2_score_round(first, second)

        print(p1_sum, p2_sum)

File: rust/src/day2.rs
use std::error::Error;
use std::fmt::{Debug, Display, Formatter};
use std::str::FromStr;
use crate::lib::Answer;

#[derive(Copy, Clone, Eq, PartialEq)]
enum Choice {
    Rock,
    Paper,
    Scissors
}

enum Outcome {
    Win,
    Lose,
    Draw
}

impl Outcome {
    fn value(&self) -> i8 {
        match self {
            Outcome::Win => 6,
            Outcome::Lose => 0,
            Outcome::Draw => 3,
        }
    }

    fn choose(&self, opp: Choice) -> Choice {
        use Outcome::*;
        use Choice::*;
        match (self, opp) {
            (Win, Scissors) => Rock,
            (Win, Rock) => Paper,
            (Win, Paper) => Scissors,
            (Lose, Rock) => Scissors,
            (Lose, Paper) => Rock,
            (Lose, Scissors) => Paper,
            (Draw, _) => opp
        }
    }
}

#[derive(Debug)]
struct ParseOutcomeError;

impl Display for ParseOutcomeError {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        write!(f, "unknown outcome")
    }
}

impl Error for ParseOutcomeError {}

impl FromStr for Outcome {
    type Err = ParseOutcomeError;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s {
            "X" => Ok(Outcome::Lose),
            "Y" => Ok(Outcome::Draw),
            "Z" => Ok(Outcome::Win),
            _ => Err(ParseOutcomeError)
        }
    }
}

#[derive(Debug)]
struct ParseChoiceError;

impl Display for ParseChoiceError {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        write!(f, "unknown choice")
    }
}

impl Error for ParseChoiceError {}

impl Choice {
    fn from_opp_str(s: &str) -> Result<Self, ParseChoiceError> {
        match s {
            "A" => Ok(Choice::Rock),
            "B" => Ok(Choice::Paper),
            "C" => Ok(Choice::Scissors),
            _ => Err(ParseChoiceError)
        }
    }

    fn from_my_str(s: &str) -> Result<Self, ParseChoiceError> {
        match s {
            "X" => Ok(Choice::Rock),
            "Y" => Ok(Choice::Paper),
            "Z" => Ok(Choice::Scissors),
            _ => Err(ParseChoiceError)
        }
    }

    fn value(&self) -> i8 {
        match self {
            Choice::Rock => 1,
            Choice::Paper => 2,
            Choice::Scissors => 3,
        }
    }

    fn beats(&self) -> Choice {
        match self {
            Choice::Rock => Choice::Scissors,
            Choice::Paper => Choice::Rock,
            Choice::Scissors => Choice::Paper
        }
    }

    fn outcome(&self, opp: Choice) -> Outcome {
        if self.beats() == opp {
            Outcome::Win
        } else if opp.beats() == *self {
            Outcome::Lose
        } else {
            Outcome::Draw
        }
    }

    fn score(&self, opp: Choice) -> i8 {
        let outcome = self.outcome(opp);
        self.value() + outcome.value()
    }
}

struct Part1Round {
    my_choice: Choice,
    opp_choice: Choice
}

impl Part1Round {
    fn score(&self) -> i8 {
        self.my_choice.score(self.opp_choice)
    }
}

#[derive(Debug)]
enum ParseRoundError {
    UnknownOppChoice,
    UnknownMyChoice,
    NotEnoughElements,
}

impl Display for ParseRoundError {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        match self {
            ParseRoundError::UnknownOppChoice => write!(f, "unknown opponent choice"),
            ParseRoundError::UnknownMyChoice => write!(f, "unknown my choice"),
            ParseRoundError::NotEnoughElements => write!(f, "not enough space-separated values on line")
        }
    }
}

impl Error for ParseRoundError {}

impl FromStr for Part1Round {
    type Err = ParseRoundError;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let mut split = s.split_whitespace();
        let opp = split.next().ok_or(ParseRoundError::NotEnoughElements)?;
        let my = split.next().ok_or(ParseRoundError::NotEnoughElements)?;
        let opp_choice = Choice::from_opp_str(opp).map_err(|_| ParseRoundError::UnknownOppChoice)?;
        let my_choice = Choice::from_my_str(my).map_err(|_| ParseRoundError::UnknownMyChoice)?;
        Ok(Self {my_choice, opp_choice})
    }
}

const INP: &str = include_str!("../../day2_input.txt");

fn from_string(s: &str) -> Result<Vec<Part1Round>, ParseRoundError> {
    s.trim().lines().map(&str::parse::<Part1Round>).collect()
}

fn part1(rounds: &[Part1Round]) -> i64 {
    rounds.iter()
        .map(|round| round.score() as i64).sum()
}

struct Part2Round {
    opp_choice: Choice,
    desired: Outcome
}

impl Part2Round {
    fn score(&self) -> i8 {
        let my_choice = self.desired.choose(self.opp_choice);
        my_choice.score(self.opp_choice)
    }
}

/// Convert a round serialized into a Part1Round into a Part2Round.
fn convert_round(round: &Part1Round) -> Part2Round {
    let desired = match round.my_choice {
        Choice::Rock => Outcome::Lose,
        Choice::Paper => Outcome::Draw,
        Choice::Scissors => Outcome::Win,
    };
    Part2Round { opp_choice: round.opp_choice, desired }
}

fn part2(rounds: &[Part2Round]) -> i64 {
    rounds.iter()
        .map(|round| round.score() as i64)
        .sum()
}

pub fn solve() -> Result<Answer, Box<dyn Error>> {
    let p1_rounds = from_string(INP)?;
    let p2_rounds: Vec<_> = p1_rounds.iter().map(convert_round).collect();
    let (p1, p2) = (part1(&p1_rounds), part2(&p2_rounds));
    Ok(Answer {part_1: p1, part_2: p2})
}